src/cdk/explorer/detectCdkProjects.ts

```diff
@@ -0,0 +1,98 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as path from 'path'
+import * as vscode from 'vscode'
+import { access, stat } from '../../shared/filesystem'
+import { readDirAsString } from '../../shared/filesystemUtilities'
+
+export interface CdkProject {
+    workspaceFolder: vscode.WorkspaceFolder
+    cdkJsonPath: string
+    treePath: string
+}
+
+export async function detectCdkProjects(
+    workspaceFolders: vscode.WorkspaceFolder[] | undefined = vscode.workspace.workspaceFolders
+): Promise<CdkProject[]> {
+    if (!workspaceFolders) {
+        return []
+    }
+
+    return (await Promise.all(workspaceFolders.map(detectCdkProjectsFromWorkspaceFolder))).reduce(
+        (accumulator: CdkProject[], current: CdkProject[]) => {
+            accumulator.push(...current)
+
+            return accumulator
+        },
+        []
+    )
+}
+
+async function detectCdkProjectsFromWorkspaceFolder(workspaceFolder: vscode.WorkspaceFolder): Promise<CdkProject[]> {
+    const result = []
+
+    for await (const treeJson of detectLocalCdkProjects({ workspaceUris: [workspaceFolder.uri] })) {
+        const cdkJsonPath = path.join(treeJson.fsPath, '../', 'cdk.json')
+        const project = { workspaceFolder: workspaceFolder, cdkJsonPath: cdkJsonPath, treePath: treeJson.fsPath }
+        result.push(project)
+    }
+
+    return result
+}
+
+export interface DetectCdkProjectsContext {
+    access: typeof access
+    readDir: typeof readDirAsString
+    stat: typeof stat
+}
+
+class DefaultDetectCdkProjectsContext implements DetectCdkProjectsContext {
+    public readonly access = access
+    public readonly readDir = readDirAsString
+    public readonly stat = stat
+}
+
+export async function* detectLocalCdkProjects({
+    workspaceUris,
+    context = new DefaultDetectCdkProjectsContext()
+}: {
+    workspaceUris: vscode.Uri[]
+    context?: DetectCdkProjectsContext
+}): AsyncIterableIterator<vscode.Uri> {
+    for (const workspaceFolder of workspaceUris) {
+        for await (const folder of getFolderCandidates(context, workspaceFolder)) {
+            yield* detectCdkProjectsInFolder(context, folder)
+        }
+    }
+}
+
+async function* getFolderCandidates(context: DetectCdkProjectsContext, uri: vscode.Uri): AsyncIterableIterator<string> {
+    // Search the root and first level of children only.
```

Comment:
We're getting a fair amount of feedback indicating this is not sufficient for detecting SAM Applications and Lambda handlers. What would the performance be if we searched the full subfolder tree? An alternative would be to make the depth traversed user-configurable.

Reply:
discussed offline - a project at the root or one layer in is more closely aligned with the natural CDK authoring workflow. Having said that, we will still revisit this topic as it would be handy to be able to find all CDK projects regardless of how deep it is into the workspace.